[Bug]: Search memory failed with Qwen3.5-xxx thinking mode (default enabled)

### Describe the bug

Thinking mode is enabled by default since Qwen3.5-xxx (qwen3.5-plus, qwen3.5-flash) and memmachine search memory could not handle <think> </think> section in Pydantic model_validate_json()

```python
2026-03-25 03:05:22,890 [ERROR] memmachine_server.semantic_memory.semantic_ingestion - Failed to process message 5 for semantic type profile
Traceback (most recent call last):
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_ingestion.py", line 163, in process_semantic_type
    commands = await llm_feature_update(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/_internal/_validate_call.py", line 34, in wrapper_function
    return await wrapper(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_llm.py", line 88, in llm_feature_update
    parsed_output = await model.generate_parsed_response(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/common/language_model/openai_responses_language_model.py", line 157, in generate_parsed_response
    response = await self._client.with_options(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/responses/responses.py", line 2825, in parse
    return await self._post(
           ^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1884, in post
    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1674, in request
    return await self._process_response(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1763, in _process_response
    return await api_response.parse()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_response.py", line 432, in parse
    parsed = self._options.post_parser(parsed)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/responses/responses.py", line 2819, in parser
    return parse_response(
           ^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_responses.py", line 74, in parse_response
    "parsed": parse_text(item.text, text_format=text_format),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_responses.py", line 139, in parse_text
    return cast(TextFormatT, model_parse_json(text_format, text))
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_compat.py", line 171, in model_parse_json
    return model.model_validate_json(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/main.py", line 766, in model_validate_json
    return cls.__pydantic_validator__.validate_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for _SemanticFeatureUpdateRes
  Invalid JSON: expected value at line 1 column 1 [type=json_invalid, input_value='<think>\nThe user is sha...atabase work"\n    }\n]', input_type=str]
    For further information visit https://errors.pydantic.dev/2.12/v/json_invalid
2026-03-25 03:05:45,040 [INFO] httpx - HTTP Request: POST https://dashscope.aliyuncs.com/compatible-mode/v1/responses "HTTP/1.1 200 OK"
2026-03-25 03:05:45,041 [ERROR] memmachine_server.semantic_memory.semantic_ingestion - Failed to process message 6 for semantic type profile
Traceback (most recent call last):
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_ingestion.py", line 163, in process_semantic_type
    commands = await llm_feature_update(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/_internal/_validate_call.py", line 34, in wrapper_function
    return await wrapper(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_llm.py", line 88, in llm_feature_update
    parsed_output = await model.generate_parsed_response(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/common/language_model/openai_responses_language_model.py", line 157, in generate_parsed_response
    response = await self._client.with_options(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/responses/responses.py", line 2825, in parse
    return await self._post(
           ^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1884, in post
    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1674, in request
    return await self._process_response(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1763, in _process_response
    return await api_response.parse()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_response.py", line 432, in parse
    parsed = self._options.post_parser(parsed)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/responses/responses.py", line 2819, in parser
    return parse_response(
           ^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_responses.py", line 74, in parse_response
    "parsed": parse_text(item.text, text_format=text_format),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_responses.py", line 139, in parse_text
    return cast(TextFormatT, model_parse_json(text_format, text))
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_compat.py", line 171, in model_parse_json
    return model.model_validate_json(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/main.py", line 766, in model_validate_json
    return cls.__pydantic_validator__.validate_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for _SemanticFeatureUpdateRes
  Invalid JSON: expected value at line 1 column 1 [type=json_invalid, input_value='<think>\nLooking at this...Third Board)"\n    }\n]', input_type=str]
    For further information visit https://errors.pydantic.dev/2.12/v/json_invalid

``` 

### Steps to reproduce

From Qwen official document, Thinking mode is enabled by default since qwen3.5

https://www.alibabacloud.com/help/en/model-studio/deep-thinking#19a4c1988fp6d

### Expected behavior

Search memory may report such error and add memory works well.

### Environment

MemMachine version 0.3.2

### Additional context

_No response_

When I reproduce this issue in local lab environment, I get another similar errors. This means we have issue to parse Qwen3.5 response.

```python
2026-03-26 06:00:30,508 [INFO] httpx - HTTP Request: POST https://dashscope.aliyuncs.com/compatible-mode/v1/chat/completions "HTTP/1.1 200 OK"
2026-03-26 06:00:30,511 [ERROR] memmachine_server.semantic_memory.semantic_ingestion - Failed to process message 46 for semantic type profile
Traceback (most recent call last):
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_ingestion.py", line 163, in process_semantic_type
    commands = await llm_feature_update(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/_internal/_validate_call.py", line 34, in wrapper_function
    return await wrapper(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_llm.py", line 88, in llm_feature_update
    parsed_output = await model.generate_parsed_response(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/common/language_model/openai_chat_completions_language_model.py", line 130, in generate_parsed_response
    response = await self._client.with_options(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/chat/completions/completions.py", line 1668, in parse
    return await self._post(
           ^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1884, in post
    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1674, in request
    return await self._process_response(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1763, in _process_response
    return await api_response.parse()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_response.py", line 432, in parse
    parsed = self._options.post_parser(parsed)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/chat/completions/completions.py", line 1662, in parser
    return _parse_chat_completion(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_completions.py", line 146, in parse_chat_completion
    "parsed": maybe_parse_content(
              ^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_completions.py", line 196, in maybe_parse_content
    return _parse_content(response_format, message.content)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_completions.py", line 245, in _parse_content
    return cast(ResponseFormatT, model_parse_json(response_format, content))
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_compat.py", line 171, in model_parse_json
    return model.model_validate_json(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/main.py", line 766, in model_validate_json
    return cls.__pydantic_validator__.validate_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for _SemanticFeatureUpdateRes
  Input should be an object [type=model_type, input_value=[], input_type=list]
    For further information visit https://errors.pydantic.dev/2.12/v/model_type
2026-03-26 06:00:36,661 [INFO] httpx - HTTP Request: POST https://dashscope.aliyuncs.com/compatible-mode/v1/chat/completions "HTTP/1.1 200 OK"
2026-03-26 06:00:36,664 [ERROR] memmachine_server.semantic_memory.semantic_ingestion - Failed to process message 48 for semantic type profile
Traceback (most recent call last):
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_ingestion.py", line 163, in process_semantic_type
    commands = await llm_feature_update(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/_internal/_validate_call.py", line 34, in wrapper_function
    return await wrapper(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/semantic_memory/semantic_llm.py", line 88, in llm_feature_update
    parsed_output = await model.generate_parsed_response(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/memmachine_server/common/language_model/openai_chat_completions_language_model.py", line 130, in generate_parsed_response
    response = await self._client.with_options(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/chat/completions/completions.py", line 1668, in parse
    return await self._post(
           ^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1884, in post
    return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1674, in request
    return await self._process_response(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1763, in _process_response
    return await api_response.parse()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_response.py", line 432, in parse
    parsed = self._options.post_parser(parsed)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/resources/chat/completions/completions.py", line 1662, in parser
    return _parse_chat_completion(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_completions.py", line 146, in parse_chat_completion
    "parsed": maybe_parse_content(
              ^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_completions.py", line 196, in maybe_parse_content
    return _parse_content(response_format, message.content)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/lib/_parsing/_completions.py", line 245, in _parse_content
    return cast(ResponseFormatT, model_parse_json(response_format, content))
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/openai/_compat.py", line 171, in model_parse_json
    return model.model_validate_json(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.venv/lib/python3.12/site-packages/pydantic/main.py", line 766, in model_validate_json
    return cls.__pydantic_validator__.validate_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for _SemanticFeatureUpdateRes
  Input should be an object [type=model_type, input_value=[], input_type=list]
    For further information visit https://errors.pydantic.dev/2.12/v/model_type

```

--- a/packages/server/server_tests/memmachine_server/common/language_model/test_openai_chat_completions_language_model.py
+++ b/packages/server/server_tests/memmachine_server/common/language_model/test_openai_chat_completions_language_model.py
@@ -1,5 +1,7 @@
 """Unit tests for OpenAIChatCompletionsLanguageModel."""
 
+from dataclasses import dataclass
+from typing import Any, cast
 from unittest.mock import AsyncMock, MagicMock, patch
 
 import httpx
@@ -9,7 +11,7 @@ from openai.types import chat as openai_chat
 from openai.types.chat.chat_completion_message_function_tool_call import (
     Function as ToolCallFunction,
 )
-from pydantic import ValidationError
+from pydantic import BaseModel, ValidationError
 
 from memmachine_server.common.data_types import ExternalServiceAPIError
 from memmachine_server.common.language_model.openai_chat_completions_language_model import (
@@ -17,6 +19,38 @@ from memmachine_server.common.language_model.openai_chat_completions_language_mo
     OpenAIChatCompletionsLanguageModelParams,
 )
 from memmachine_server.common.metrics_factory import MetricsFactory
+
+
+class FakeAsyncStream:
+    def __init__(self, chunks):
+        self._chunks = iter(chunks)
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        try:
+            return next(self._chunks)
+        except StopIteration as exc:
+            raise StopAsyncIteration from exc
+
+
+@dataclass
+class FakeReasoningChunk:
+    type: str = "response.reasoning.delta"
+
+
+class ParsedResponse(BaseModel):
+    answer: str
+    confidence: float
+
+
+def make_chat_completion(data: dict[str, Any]) -> openai_chat.ChatCompletion:
+    return openai_chat.ChatCompletion.model_validate(data)
+
+
+def make_chat_completion_chunk(data: dict[str, Any]) -> openai_chat.ChatCompletionChunk:
+    return openai_chat.ChatCompletionChunk.model_validate(data)
 
 
 @pytest.fixture
@@ -214,6 +248,338 @@ async def test_generate_response_success(mock_async_openai, minimal_config):
         {"role": "user", "content": "User prompt"},
     ]
     assert call_args.kwargs["store"] is False
+
+
+@pytest.mark.asyncio
+async def test_generate_parsed_response_success(mock_async_openai, minimal_config):
+    """Test a non-streamed structured response is parsed correctly."""
+    mock_response = make_chat_completion(
+        {
+            "id": "completion-1",
+            "object": "chat.completion",
+            "created": 1,
+            "model": "test-model",
+            "choices": [
+                {
+                    "index": 0,
+                    "finish_reason": "stop",
+                    "message": {
+                        "role": "assistant",
+                        "content": '{"answer":"ok","confidence":0.9}',
+                        "refusal": None,
+                    },
+                }
+            ],
+            "usage": {
+                "prompt_tokens": 7,
+                "completion_tokens": 5,
+                "total_tokens": 12,
+            },
+        }
+    )
+
+    mock_client = mock_async_openai.return_value
+    mock_client.chat.completions.create.return_value = mock_response
+
+    lm = OpenAIChatCompletionsLanguageModel(minimal_config)
+    parsed = await lm.generate_parsed_response(
+        ParsedResponse,
+        system_prompt="System prompt",
+        user_prompt="User prompt",
+    )
+
+    assert parsed == ParsedResponse(answer="ok", confidence=0.9)
+    call_args = mock_client.chat.completions.create.call_args
+    assert call_args.kwargs["model"] == "test-model"
+    assert call_args.kwargs["messages"] == [
+        {"role": "system", "content": "System prompt"},
+        {"role": "user", "content": "User prompt"},
+    ]
+    assert call_args.kwargs["response_format"]["type"] == "json_schema"
+
+
+@pytest.mark.asyncio
+async def test_generate_parsed_response_from_stream(
+    mock_async_openai,
+    minimal_config,
+):
+    """Test a streamed structured response is aggregated and validated."""
+    streamed_response = FakeAsyncStream(
+        [
+            FakeReasoningChunk(),
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-1",
+                    "object": "chat.completion.chunk",
+                    "created": 1,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {
+                                "role": "assistant",
+                                "content": '{"answer":"ok",',
+                            },
+                            "finish_reason": None,
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": None,
+                }
+            ),
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-2",
+                    "object": "chat.completion.chunk",
+                    "created": 2,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {"content": '"confidence":0.9}'},
+                            "finish_reason": "stop",
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": None,
+                }
+            ),
+        ]
+    )
+
+    mock_client = mock_async_openai.return_value
+    mock_client.chat.completions.create.return_value = streamed_response
+
+    lm = OpenAIChatCompletionsLanguageModel(minimal_config)
+    parsed = await lm.generate_parsed_response(ParsedResponse)
+
+    assert parsed == ParsedResponse(answer="ok", confidence=0.9)
+
+
+@pytest.mark.asyncio
+async def test_generate_parsed_response_from_stream_discards_reasoning_delta_chunks(
+    mock_async_openai,
+    minimal_config,
+):
+    """Test streamed structured responses ignore chunks with reasoning_content."""
+    reasoning_chunk = make_chat_completion_chunk(
+        {
+            "id": "chunk-0",
+            "object": "chat.completion.chunk",
+            "created": 0,
+            "model": "test-model",
+            "choices": [
+                {
+                    "index": 0,
+                    "delta": {"content": None},
+                    "finish_reason": None,
+                    "logprobs": None,
+                }
+            ],
+            "usage": None,
+        }
+    )
+    cast(Any, reasoning_chunk.choices[0].delta).reasoning_content = "chain-of-thought"
+
+    streamed_response = FakeAsyncStream(
+        [
+            reasoning_chunk,
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-1",
+                    "object": "chat.completion.chunk",
+                    "created": 1,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {
+                                "role": "assistant",
+                                "content": '{"answer":"ok",',
+                            },
+                            "finish_reason": None,
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": None,
+                }
+            ),
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-2",
+                    "object": "chat.completion.chunk",
+                    "created": 2,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {"content": '"confidence":0.9}'},
+                            "finish_reason": "stop",
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": None,
+                }
+            ),
+        ]
+    )
+
+    mock_client = mock_async_openai.return_value
+    mock_client.chat.completions.create.return_value = streamed_response
+
+    lm = OpenAIChatCompletionsLanguageModel(minimal_config)
+    parsed = await lm.generate_parsed_response(ParsedResponse)
+
+    assert parsed == ParsedResponse(answer="ok", confidence=0.9)
+
+
+@pytest.mark.asyncio
+async def test_generate_response_streamed_chat_completion_chunks(
+    mock_async_openai,
+    minimal_config,
+):
+    """Test a streamed chat completion is aggregated into a single response."""
+    streamed_response = FakeAsyncStream(
+        [
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-1",
+                    "object": "chat.completion.chunk",
+                    "created": 1,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {"role": "assistant", "content": "Hello"},
+                            "finish_reason": None,
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": None,
+                }
+            ),
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-2",
+                    "object": "chat.completion.chunk",
+                    "created": 2,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {"content": ", world!"},
+                            "finish_reason": "stop",
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": {
+                        "prompt_tokens": 10,
+                        "completion_tokens": 4,
+                        "total_tokens": 14,
+                    },
+                }
+            ),
+        ]
+    )
+
+    mock_client = mock_async_openai.return_value
+    mock_client.chat.completions.create.return_value = streamed_response
+
+    lm = OpenAIChatCompletionsLanguageModel(minimal_config)
+    (
+        content,
+        tool_calls,
+        input_tokens,
+        output_tokens,
+    ) = await lm.generate_response_with_token_usage()
+
+    assert content == "Hello, world!"
+    assert tool_calls == []
+    assert input_tokens == 10
+    assert output_tokens == 4
+
+
+@pytest.mark.asyncio
+async def test_generate_response_streamed_tool_calls_and_discards_reasoning_chunks(
+    mock_async_openai,
+    minimal_config,
+):
+    """Test streamed tool calls are reconstructed while reasoning chunks are ignored."""
+    streamed_response = FakeAsyncStream(
+        [
+            FakeReasoningChunk(),
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-1",
+                    "object": "chat.completion.chunk",
+                    "created": 1,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {
+                                "tool_calls": [
+                                    {
+                                        "index": 0,
+                                        "id": "call_123",
+                                        "type": "function",
+                                        "function": {
+                                            "name": "get_weather",
+                                            "arguments": '{"location"',
+                                        },
+                                    }
+                                ]
+                            },
+                            "finish_reason": None,
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": None,
+                }
+            ),
+            make_chat_completion_chunk(
+                {
+                    "id": "chunk-2",
+                    "object": "chat.completion.chunk",
+                    "created": 2,
+                    "model": "test-model",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {
+                                "tool_calls": [
+                                    {
+                                        "index": 0,
+                                        "function": {"arguments": ': "Boston"}'},
+                                    }
+                                ]
+                            },
+                            "finish_reason": "tool_calls",
+                            "logprobs": None,
+                        }
+                    ],
+                    "usage": None,
+                }
+            ),
+        ]
+    )
+
+    mock_client = mock_async_openai.return_value
+    mock_client.chat.completions.create.return_value = streamed_response
+
+    lm = OpenAIChatCompletionsLanguageModel(minimal_config)
+    content, tool_calls = await lm.generate_response()
+
+    assert content == ""
+    assert tool_calls == [
+        {
+            "call_id": "call_123",
+            "function": {
+                "name": "get_weather",
+                "arguments": {"location": "Boston"},
+            },
+        }
+    ]
 
 
 @pytest.mark.asyncio

--- a/packages/server/src/memmachine_server/common/language_model/openai_chat_completions_language_model.py
+++ b/packages/server/src/memmachine_server/common/language_model/openai_chat_completions_language_model.py
@@ -2,13 +2,16 @@
 
 import asyncio
 import logging
+from collections.abc import AsyncIterator
 from typing import Any, TypeVar, cast
 from uuid import uuid4
 
 import json_repair
 import openai
+from openai.lib._parsing import parse_chat_completion, type_to_response_format_param
 from openai.types.chat import (
     ChatCompletion,
+    ChatCompletionChunk,
     ChatCompletionMessageFunctionToolCall,
 )
 from pydantic import BaseModel, Field, InstanceOf, TypeAdapter
@@ -21,6 +24,13 @@ from .language_model import LanguageModel
 T = TypeVar("T")
 
 logger = logging.getLogger(__name__)
+
+
+class _StreamedToolCallState(BaseModel):
+    id: str | None = None
+    type: str | None = None
+    function_name: str = ""
+    function_arguments: str = ""
 
 
 class OpenAIChatCompletionsLanguageModelParams(BaseModel):
@@ -126,28 +136,20 @@ class OpenAIChatCompletionsLanguageModel(LanguageModel):
 
             generate_response_call_uuid = uuid4()
 
-            try:
-                response = await self._client.with_options(
-                    max_retries=max_attempts,
-                ).chat.completions.parse(
-                    model=self._model,
-                    messages=input_prompts,
-                    response_format=output_format,
-                    store=False,
-                )
-            except openai.OpenAIError as e:
-                error_message = (
-                    f"[call uuid: {generate_response_call_uuid}] "
-                    "Giving up generating response "
-                    f"due to non-retryable {type(e).__name__}"
-                )
-                logger.exception(error_message)
-                raise ExternalServiceAPIError(error_message) from e
+            response = await self._request_chat_completion(
+                args={
+                    "model": self._model,
+                    "messages": input_prompts,
+                    "response_format": type_to_response_format_param(output_format),
+                    "store": False,
+                },
+                max_attempts=max_attempts,
+                generate_response_call_uuid=generate_response_call_uuid,
+            )
 
-            self._collect_usage_metrics(response)
-
-            return TypeAdapter(output_format).validate_python(
-                response.choices[0].message.parsed
+            return await self._normalize_parsed_chat_completion_response(
+                response=response,
+                output_format=output_format,
             )
 
     async def generate_response(
@@ -183,7 +185,7 @@ class OpenAIChatCompletionsLanguageModel(LanguageModel):
             max_attempts=max_attempts,
         )
 
-    async def _generate_response(  # noqa: C901
+    async def _generate_response(
         self,
         system_prompt: str | None = None,
         user_prompt: str | None = None,
@@ -205,99 +207,74 @@ class OpenAIChatCompletionsLanguageModel(LanguageModel):
             )
             generate_response_call_uuid = uuid4()
 
-            sleep_seconds = 1
-            for attempt in range(1, max_attempts + 1):
-                try:
-                    args: dict = {
-                        "model": self._model,
-                        "messages": input_prompts,
-                        "store": False,
-                    }
-                    if tools:
-                        args["tools"] = tools
-                        args["tool_choice"] = (
-                            tool_choice if tool_choice is not None else "auto"
-                        )
-                    response = await self._client.chat.completions.create(**args)
-                    break
-                except (
-                    openai.RateLimitError,
-                    openai.APITimeoutError,
-                    openai.APIConnectionError,
-                    openai.InternalServerError,
-                ) as e:
-                    # Exception may be retried.
-                    if attempt >= max_attempts:
-                        error_message = (
-                            f"[call uuid: {generate_response_call_uuid}] "
-                            "Giving up generating response "
-                            f"after failed attempt {attempt} "
-                            f"due to retryable {type(e).__name__}: "
-                            f"max attempts {max_attempts} reached"
-                        )
-                        logger.exception(error_message)
-                        raise ExternalServiceAPIError(error_message) from e
+            args: dict[str, Any] = {
+                "model": self._model,
+                "messages": input_prompts,
+                "store": False,
+            }
+            if tools:
+                args["tools"] = tools
+                args["tool_choice"] = tool_choice if tool_choice is not None else "auto"
 
-                    logger.info(
-                        "[call uuid: %s] "
-                        "Retrying generating response in %d seconds "
-                        "after failed attempt %d due to retryable %s...",
-                        generate_response_call_uuid,
-                        sleep_seconds,
-                        attempt,
-                        type(e).__name__,
-                    )
-                    await asyncio.sleep(sleep_seconds)
-                    sleep_seconds *= 2
-                    sleep_seconds = min(sleep_seconds, self._max_retry_interval_seconds)
-                    continue
-                except openai.OpenAIError as e:
+            response = await self._request_chat_completion(
+                args=args,
+                max_attempts=max_attempts,
+                generate_response_call_uuid=generate_response_call_uuid,
+            )
+
+            return await self._normalize_chat_completion_response(response)
+
+    async def _request_chat_completion(
+        self,
+        args: dict[str, Any],
+        max_attempts: int,
+        generate_response_call_uuid: object,
+    ) -> ChatCompletion | AsyncIterator[object] | object:
+        sleep_seconds = 1
+        for attempt in range(1, max_attempts + 1):
+            try:
+                return await self._client.chat.completions.create(**args)
+            except (
+                openai.RateLimitError,
+                openai.APITimeoutError,
+                openai.APIConnectionError,
+                openai.InternalServerError,
+            ) as e:
+                if attempt >= max_attempts:
                     error_message = (
                         f"[call uuid: {generate_response_call_uuid}] "
                         "Giving up generating response "
                         f"after failed attempt {attempt} "
-                        f"due to non-retryable {type(e).__name__}"
+                        f"due to retryable {type(e).__name__}: "
+                        f"max attempts {max_attempts} reached"
                     )
                     logger.exception(error_message)
                     raise ExternalServiceAPIError(error_message) from e
 
-            self._collect_usage_metrics(response)
+                logger.info(
+                    "[call uuid: %s] "
+                    "Retrying generating response in %d seconds "
+                    "after failed attempt %d due to retryable %s...",
+                    generate_response_call_uuid,
+                    sleep_seconds,
+                    attempt,
+                    type(e).__name__,
+                )
+                await asyncio.sleep(sleep_seconds)
+                sleep_seconds *= 2
+                sleep_seconds = min(sleep_seconds, self._max_retry_interval_seconds)
+                continue
+            except openai.OpenAIError as e:
+                error_message = (
+                    f"[call uuid: {generate_response_call_uuid}] "
+                    "Giving up generating response "
+                    f"after failed attempt {attempt} "
+                    f"due to non-retryable {type(e).__name__}"
+                )
+                logger.exception(error_message)
+                raise ExternalServiceAPIError(error_message) from e
 
-            function_calls_arguments = []
-            try:
-                if response.choices[0].message.tool_calls:
-                    for tool_call in response.choices[0].message.tool_calls:
-                        if isinstance(
-                            tool_call,
-                            ChatCompletionMessageFunctionToolCall,
-                        ):
-                            function_calls_arguments.append(
-                                {
-                                    "call_id": tool_call.id,
-                                    "function": {
-                                        "name": tool_call.function.name,
-                                        "arguments": json_repair.loads(
-                                            tool_call.function.arguments,
-                                        ),
-                                    },
-                                },
-                            )
-                        else:
-                            logger.info(
-                                "Unsupported tool call type: %s",
-                                type(tool_call).__name__,
-                            )
-            except (TypeError, ValueError) as e:
-                raise ValueError(
-                    "Failed to repair or parse JSON from function call arguments"
-                ) from e
-
-            return (
-                response.choices[0].message.content or "",
-                function_calls_arguments,
-                response.usage.prompt_tokens if response.usage else 0,
-                response.usage.completion_tokens if response.usage else 0,
-            )
+        raise RuntimeError("unreachable")
 
     def _collect_usage_metrics(self, response: ChatCompletion) -> None:
         if not self._should_collect_metrics:
@@ -317,5 +294,255 @@ class OpenAIChatCompletionsLanguageModel(LanguageModel):
             self._total_tokens_usage_counter.increment(
                 value=response.usage.total_tokens,
             )
+        except Exception:
+            logger.exception("Failed to collect usage metrics")
+
+    async def _normalize_chat_completion_response(
+        self,
+        response: ChatCompletion | AsyncIterator[object] | object,
+    ) -> tuple[str, list[dict[str, Any]], int, int]:
+        if not self._is_streaming_response(response):
+            response = cast(ChatCompletion, response)
+            self._collect_usage_metrics(response)
+            function_calls_arguments = self._extract_tool_calls_from_completion(
+                response
+            )
+            return (
+                response.choices[0].message.content or "",
+                function_calls_arguments,
+                response.usage.prompt_tokens if response.usage else 0,
+                response.usage.completion_tokens if response.usage else 0,
+            )
+
+        return await self._collect_streaming_response(cast(Any, response))
+
+    async def _normalize_parsed_chat_completion_response(
+        self,
+        response: ChatCompletion | AsyncIterator[object] | object,
+        output_format: type[T],
+    ) -> T:
+        if not self._is_streaming_response(response):
+            response = cast(ChatCompletion, response)
+            self._collect_usage_metrics(response)
+            parsed_response = parse_chat_completion(
+                response_format=output_format,
+                input_tools=[],
+                chat_completion=response,
+            )
+            return TypeAdapter(output_format).validate_python(
+                parsed_response.choices[0].message.parsed
+            )
+
+        content, _, _, _ = await self._collect_streaming_response(cast(Any, response))
+        return TypeAdapter(output_format).validate_python(json_repair.loads(content))
+
+    def _extract_tool_calls_from_completion(
+        self,
+        response: ChatCompletion,
+    ) -> list[dict[str, Any]]:
+        function_calls_arguments = []
+        try:
+            if response.choices[0].message.tool_calls:
+                for tool_call in response.choices[0].message.tool_calls:
+                    if isinstance(
+                        tool_call,
+                        ChatCompletionMessageFunctionToolCall,
+                    ):
+                        function_calls_arguments.append(
+                            {
+                                "call_id": tool_call.id,
+                                "function": {
+                                    "name": tool_call.function.name,
+                                    "arguments": json_repair.loads(
+                                        tool_call.function.arguments,
+                                    ),
+                                },
+                            },
+                        )
+                    else:
+                        logger.info(
+                            "Unsupported tool call type: %s",
+                            type(tool_call).__name__,
+                        )
+        except (TypeError, ValueError) as e:
+            raise ValueError(
+                "Failed to repair or parse JSON from function call arguments"
+            ) from e
+
+        return function_calls_arguments
+
+    async def _collect_streaming_response(
+        self,
+        response_stream: AsyncIterator[object],
+    ) -> tuple[str, list[dict[str, Any]], int, int]:
+        output_chunks: list[str] = []
+        streamed_tool_calls: dict[int, _StreamedToolCallState] = {}
+        prompt_tokens = 0
+        completion_tokens = 0
+        total_tokens = 0
+
+        async for chunk in response_stream:
+            if not self._is_supported_stream_chunk(chunk):
+                continue
+
+            chat_chunk = cast(ChatCompletionChunk, chunk)
+            prompt_tokens, completion_tokens, total_tokens = (
+                self._update_stream_usage_counters(
+                    chat_chunk,
+                    prompt_tokens,
+                    completion_tokens,
+                    total_tokens,
+                )
+            )
+            self._append_stream_chunk_data(
+                chat_chunk,
+                output_chunks,
+                streamed_tool_calls,
+            )
+
+        self._collect_stream_usage_metrics(
+            prompt_tokens,
+            completion_tokens,
+            total_tokens,
+        )
+
+        function_calls_arguments = self._build_streamed_tool_calls(streamed_tool_calls)
+
+        return (
+            "".join(output_chunks),
+            function_calls_arguments,
+            prompt_tokens,
+            completion_tokens,
+        )
+
+    def _build_streamed_tool_calls(
+        self,
+        streamed_tool_calls: dict[int, _StreamedToolCallState],
+    ) -> list[dict[str, Any]]:
+        function_calls_arguments = []
+
+        try:
+            for index in sorted(streamed_tool_calls):
+                tool_call = streamed_tool_calls[index]
+                if tool_call.type != "function":
+                    logger.info(
+                        "Unsupported streamed tool call type: %s",
+                        tool_call.type,
+                    )
+                    continue
+                function_calls_arguments.append(
+                    {
+                        "call_id": tool_call.id,
+                        "function": {
+                            "name": tool_call.function_name,
+                            "arguments": json_repair.loads(
+                                tool_call.function_arguments,
+                            ),
+                        },
+                    },
+                )
+        except (TypeError, ValueError) as e:
+            raise ValueError(
+                "Failed to repair or parse JSON from function call arguments"
+            ) from e
+
+        return function_calls_arguments
+
+    def _is_reasoning_chunk(self, chunk: object) -> bool:
+        chunk_type = getattr(chunk, "type", None)
+        if isinstance(chunk_type, str) and "reason" in chunk_type.lower():
+            logger.debug("Discarding streamed reasoning item of type %s", chunk_type)
+            return True
+
+        if isinstance(chunk, ChatCompletionChunk) and chunk.choices:
+            delta = chunk.choices[0].delta
+            if hasattr(delta, "reasoning_content"):
+                logger.debug(
+                    "Discarding streamed reasoning chunk with reasoning content"
+                )
+                return True
+
+        return False
+
+    def _is_streaming_response(self, response: object) -> bool:
+        if isinstance(response, ChatCompletion) or hasattr(response, "choices"):
+            return False
+        if type(response).__module__.startswith("unittest.mock"):
+            return False
+        return isinstance(response, openai.AsyncStream) or hasattr(
+            response, "__anext__"
+        )
+
+    def _is_supported_stream_chunk(self, chunk: object) -> bool:
+        if self._is_reasoning_chunk(chunk):
+            return False
+
+        if isinstance(chunk, ChatCompletionChunk):
+            return True
+
+        logger.info(
+            "Discarding unsupported streamed chat completion item: %s",
+            type(chunk).__name__,
+        )
+        return False
+
+    def _update_stream_usage_counters(
+        self,
+        chunk: ChatCompletionChunk,
+        prompt_tokens: int,
+        completion_tokens: int,
+        total_tokens: int,
+    ) -> tuple[int, int, int]:
+        if chunk.usage is None:
+            return prompt_tokens, completion_tokens, total_tokens
+        return (
+            chunk.usage.prompt_tokens + prompt_tokens,
+            chunk.usage.completion_tokens + completion_tokens,
+            chunk.usage.total_tokens + total_tokens,
+        )
+
+    def _append_stream_chunk_data(
+        self,
+        chunk: ChatCompletionChunk,
+        output_chunks: list[str],
+        streamed_tool_calls: dict[int, _StreamedToolCallState],
+    ) -> None:
+        for choice in chunk.choices:
+            delta = choice.delta
+
+            if delta.content:
+                output_chunks.append(delta.content)
+
+            if not delta.tool_calls:
+                continue
+
+            for tool_call in delta.tool_calls:
+                state = streamed_tool_calls.setdefault(
+                    tool_call.index,
+                    _StreamedToolCallState(),
+                )
+                if tool_call.id:
+                    state.id = tool_call.id
+                if tool_call.type:
+                    state.type = tool_call.type
+                if tool_call.function is not None:
+                    if tool_call.function.name:
+                        state.function_name += tool_call.function.name
+                    if tool_call.function.arguments:
+                        state.function_arguments += tool_call.function.arguments
+
+    def _collect_stream_usage_metrics(
+        self,
+        prompt_tokens: int,
+        completion_tokens: int,
+        total_tokens: int,
+    ) -> None:
+        if not self._should_collect_metrics or total_tokens <= 0:
+            return
+
+        try:
+            self._input_tokens_usage_counter.increment(value=prompt_tokens)
+            self._output_tokens_usage_counter.increment(value=completion_tokens)
+            self._total_tokens_usage_counter.increment(value=total_tokens)
         except Exception:
             logger.exception("Failed to collect usage metrics")

